confmap/provider/s3mapprovider/provider.go

```diff
@@ -0,0 +1,135 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3mapprovider // import "github.com/open-telemetry/opentelemetry-collector-contrib/confmap/provider/s3mapprovider"
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"go.opentelemetry.io/collector/confmap"
+	"gopkg.in/yaml.v2"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+)
+
+const (
+	schemeName = "s3"
+)
+
+type s3Client interface {
+	GetObject(context.Context, *s3.GetObjectInput, ...func(*s3.Options)) (*s3.GetObjectOutput, error)
+}
+
+type provider struct {
+	client s3Client
+}
+
+// New returns a new confmap.Provider that reads the configuration from a file.
+//
+// This Provider supports "s3" scheme, and can be called with a "uri" that follows:
+//   s3-uri : s3://[BUCKET].s3.[REGION].amazonaws.com/[KEY]
+//
+// One example for s3-uri be like: s3://DOC-EXAMPLE-BUCKET.s3.us-west-2.amazonaws.com/photos/puppy.jpg
+//
+// Examples:
+// `s3://DOC-EXAMPLE-BUCKET.s3.us-west-2.amazonaws.com/photos/puppy.jpg` - (unix, windows)
+func New() confmap.Provider {
+	// initialize the client
+	cfg, err := config.LoadDefaultConfig(context.Background())
+	if err != nil {
+		log.Fatal(err)
+	}
```

Comment:
We should not use `log.Fatal` here, we never use the `log` package and this error may therefore be missed since it will not appear alongside the rest of logs.

I would initialize the S3 client on the first call to `Retrieve` now, and propose allowing the `New` constructor to return an error on opentelemetry-collector.

Reply:
Fixed the comments of `log.Fatal`, and `initialize the S3 client on the first call to Retrieve now`.

But for `allowing the New constructor to return an error on opentelemetry-collector `, I think it would better start a new issue and a new PR?